Implementing COSMOS support

Jbsco opened the pull request:
Adds [README](https://github.com/lasp/adamant_example/compare/main...Jbsco:adamant_example:main#diff-3df39c183ea49623b72c7b52f97021c22876991431f5f46a935b188c6a0d7b24) instructions and information to get [OpenC3 COSMOS](https://github.com/openc3/cosmos) running with the Linux example project. This allows Rx/Tx and parsing of Adamant telemetry and command packets in COSMOS.
The added instructions detail using the plugin generation autocode added in [Adamant #21](https://github.com/lasp/adamant/pull/21) and copying the configuration to a generated plugin template in COSMOS's own [example project](https://github.com/openc3/cosmos-project).

Jbsco committed:
Update readme with COSMOS example instructions

Jbsco committed:
Reduce COSMOS setup commands to a `redo` script

dinkelk commented on gnd/cosmos/build_cosmos_plugin.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
```
I am wondering if this might be better stored with Adamant instead of the example

dinkelk commented on src/assembly/linux/main/README.md:
Cosmos -> COSMOS. That is the way they spell it.

Jbsco replied:
Changed to reflect this in [0bb46d3](https://github.com/lasp/adamant_example/pull/18/commits/0bb46d396ecaf6f5e325fb917714d9008f9738c2).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+outgoing commands.
+
+These files are autocoded by Adamant. From the `adamant_example/src/assembly/linux/main` directory in your Adamant environment run:
+
```
I think it would flow better to have users install COSMOS first.

Jbsco replied:
Agree! Order changed in [0bb46d3](https://github.com/lasp/adamant_example/pull/18/commits/0bb46d396ecaf6f5e325fb917714d9008f9738c2).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+cosmos-project/openc3-cosmos-linux-example/targets/LINUX_EXAMPLE/cmd_tlm/tlm.txt
+```
+
+Replace the contents of these text files with the contents of the corresponding files generated by Redo, located in the Adamant example project directory at:
````
I think we should create an "install_cosmos_plugin.sh" script that takes as argument the location of `cosmos-project/openc3-cosmos-linux-example/`. The script should `redo-ifchange cosmos_config` and then do the copying for the user. This has the advantage that if we expand our cosmos support over time to have more autocoded configuration files, we can just modify this script without changing the instructions for the user.

Jbsco replied:
Agreed, this is an excellent idea!

Jbsco replied:
Added a basic cp script in [bb811b5](https://github.com/lasp/adamant_example/pull/18/commits/bb811b5778578ff86f471a4b5788bb8db3ac9282).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+And must be placed in the Cosmos plugin directory at:
+
+```
+cosmos-project/openc3-cosmos-linux-example/targets/LINUX_EXAMPLE/lib/cmd_checksum.rb
````
Again the script can do this.

Jbsco replied:
Now handled by the cp script in [e9cca6c](https://github.com/lasp/adamant_example/pull/18/commits/e9cca6c54dc662b329177ac1e241fef49d2a10e7).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+$ cp adamant_example/src/assembly/linux/build/cosmos/plugin/linux_example_ccsds_cosmos_commands.txt cosmos-project/openc3-cosmos-linux-example/targets/LINUX_EXAMPLE/cmd_tlm/cmd.txt
+$ cp adamant_example/src/assembly/linux/build/cosmos/plugin/linux_example_ccsds_cosmos_telemetry.txt cosmos-project/openc3-cosmos-linux-example/targets/LINUX_EXAMPLE/cmd_tlm/tlm.txt
+$ cp adamant_example/src/assembly/linux/build/cosmos/plugin/linux_example_ccsds_cosmos_plugin.txt cosmos-project/openc3-cosmos-linux-example/plugin.txt
+$ cp adamant/gnd/cosmos/cmd_checksum.rb cosmos-project/openc3-cosmos-linux-example/targets/LINUX_EXAMPLE/lib/cmd_checksum.rb
```
Make sure we specify that this all needs to be done on the host machine, not inside the Adamant env.

Jbsco replied:
Updated to mention this, also now handled by the cp script in [7da1342](https://github.com/lasp/adamant_example/pull/18/commits/7da1342485240bf37ea3b002aa2953ecb17e7ceb).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+```
+    ports:
+      - "2003:2003/tcp"
+```
````
This is annoying. But I am not seeing a way around it...

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+      - "2003:2003/tcp"
+```
+
+With the Linux assembly running, start Cosmos by navigating to the `cosmos-project` directory and running `./openc3.sh start`. After the containers start, open a web browser and navigate to `http://localhost:2900`. You will be prompted to configure a password, and then presented with the Cosmos interface. Some modules may still be loading, but after a minute the interface should be complete. Navi ...
````
Include the command here to start the linux assembly, just so it is clear as day.

Jbsco replied:
Addressed in [e9cca6c](https://github.com/lasp/adamant_example/pull/18/commits/e9cca6c54dc662b329177ac1e241fef49d2a10e7).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+      - "2003:2003/tcp"
+```
+
+With the Linux assembly running, start Cosmos by navigating to the `cosmos-project` directory and running `./openc3.sh start`. After the containers start, open a web browser and navigate to `http://localhost:2900`. You will be prompted to configure a password, and then presented with the Cosmos interface. Some modules may still be loading, but after a minute the interface should be complete. Navi ...
````
I think all commands the user has to run explicitly really should be in 
```
$ blocks like this so it is clear
```

Jbsco replied:
Addressed in [bb811b5](https://github.com/lasp/adamant_example/pull/18/commits/bb811b5778578ff86f471a4b5788bb8db3ac9282).
Should directories and addresses also be described in blocks?

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+      - "2003:2003/tcp"
+```
+
+With the Linux assembly running, start Cosmos by navigating to the `cosmos-project` directory and running `./openc3.sh start`. After the containers start, open a web browser and navigate to `http://localhost:2900`. You will be prompted to configure a password, and then presented with the Cosmos interface. Some modules may still be loading, but after a minute the interface should be complete. Navi ...
````
Start the linux assembly after COSMOS is up and running. The order suggested here did not work on my machine.

Jbsco replied:
Order changed in [bb811b5](https://github.com/lasp/adamant_example/pull/18/commits/bb811b5778578ff86f471a4b5788bb8db3ac9282).

Jbsco committed:
Apply requested changes to readme

Jbsco committed:
Add `cp` script to handle generated COSMOS plugin files

Update readme to apply helper `cp` script

Make `cp` script executable

dinkelk commented on src/assembly/linux/main/install_cosmos_plugin.sh:
```diff
@@ -0,0 +1,17 @@
+adamant_protocol_dir=`realpath ../../../../../adamant/gnd/cosmos`
+echo cp $adamant_example_plugin_dir/linux_example_ccsds_cosmos_commands.txt $cosmos_plugin_dir/targets/LINUX_EXAMPLE/cmd_tlm/cmd.txt
+echo cp $adamant_example_plugin_dir/linux_example_ccsds_cosmos_telemetry.txt $cosmos_plugin_dir/targets/LINUX_EXAMPLE/cmd_tlm/tlm.txt
+echo cp $adamant_example_plugin_dir/linux_example_ccsds_cosmos_plugin.txt $cosmos_plugin_dir/plugin.txt
```
Hard-coding "linux_" in this script makes it not reusable for other assemblies. We should get the assembly name by looking for a .assembly.yaml file in the directory. Then we can call this script to install a plugin for any Adamant assembly. The easiest way to do this might actually be to just pass the path to the assembly file as argument to this script. One more note, you should be able to use $ADAMANT_DIR to find the root of the adamant directory. That is defined once the Adamant environment is activated.

Jbsco replied:
Addressed in [f12e5b9](https://github.com/lasp/adamant_example/pull/18/commits/f12e5b964c0e0ff3b5c7038a5858dea2276901eb). This script now takes the assembly name as an argument, Readme is also updated to reflect this.

Jbsco committed:
Update the COSMOS helper script to take the Adamant assembly & COSMOS install location as arguments

Update the COSMOS script to handle variation in plugin directory case & hyphenation

Jbsco committed:
Update the COSMOS plugin helper script to check for requested protocol files before copy

Update readme to callout COSMOS helper script handling of requested protocols

plugin helper script now copies configuration files as intended

dinkelk commented on gnd/cosmos/install_cosmos_plugin.sh:
```diff
@@ -0,0 +1,27 @@
+# Ex. `./install_cosmos_plugin.sh linux_example cosmos-project` if COSMOS is adjacent to adamant_example
+
+adamant_assembly_name=$1
+cosmos_install_name=$2
```
Add some error handling to exit out with message if either of these variables are not provided by the caller.

Jbsco replied:
Addressed in [ee6e5c3](https://github.com/lasp/adamant_example/pull/18/commits/ee6e5c38dbcb9c5ea205b8ae7f0dad1ff8f03a81).

dinkelk commented on gnd/cosmos/install_cosmos_plugin.sh:
```diff
@@ -0,0 +1,27 @@
+requested_protocol_array=($(grep \"*.rb "${adamant_example_plugin_dir}/${adamant_assembly_name}_ccsds_cosmos_plugin.txt" | cut -d' ' -f5))
+cp $adamant_example_plugin_dir/${adamant_assembly_name}_ccsds_cosmos_commands.txt $cosmos_plugin_dir/targets/${adamant_assembly_name^^}/cmd_tlm/cmd.txt
+cp $adamant_example_plugin_dir/${adamant_assembly_name}_ccsds_cosmos_telemetry.txt $cosmos_plugin_dir/targets/${adamant_assembly_name^^}/cmd_tlm/tlm.txt
+cp $adamant_example_plugin_dir/${adamant_assembly_name}_ccsds_cosmos_plugin.txt $cosmos_plugin_dir/plugin.txt
```
Why not just copy everything in this directory instead of calling things out specifically?

Jbsco replied:
The text file output from Adamant is renamed for COSMOS, the plugin file destination is also different from the cmd/tlm files.

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+A helper script, which takes the Adamant assembly name and COSMOS install directory as arguments, is provided to copy the plugin configuration files, and any custom protocols requested by the configuration, to the correct directories. If the COSMOS and Adamant example project directories are adjacent, complete the configuration by navigating to `adamant_example/gnd/cosmos` on the host machine and ...
+
+```
+$ ./install_cosmos_plugin.sh linux_example cosmos-project
````
I think using full paths here would be clearer, ie something like
```
$ adamant_example/gnd/cosmos/install_cosmos_plugin.sh adamant_example/src/assembly/linux_example /path/to/cosmos-project
```

Something like that.

Jbsco replied:
For this script the first argument is the name of the assembly, which is needed to complete parsing of the COSMOS directory structure, since it uses this name in different configurations. The second argument is a path, relative to the location of the `adamant_example` directory, should this be a full path instead of relative for the user?
Other instructions in the Readme direct the user to navigate to a directory before running a command or script, in this case should it be the full path after navigating to the top directory, e.g. `$ ./adamant_example/gnd/cosmos/install_cosmos_plugin.sh`?

dinkelk replied:
Once this script is fixed up, you should remove all the business about manually copying things above. We should just encourage users to use the script.

dinkelk replied:
I think the script should assume the directories are relative to `cwd`, ie. wherever the user is currently located, unless a full path is given. That is the way `cp` or `mv` works.

Jbsco replied:
Changed to use paths relative to CWD and update README to reflect this in [df94172](https://github.com/lasp/adamant_example/pull/18/commits/df9417225003a2134e03a62368e4e4b89f8cd4bd), plus supporting updates from https://github.com/Jbsco/adamant_example/pull/2 and https://github.com/Jbsco/adamant/pull/2.

dinkelk reviewed (changes requested):
A few changes. Please let me know when this is ready for me to try out.

Jbsco committed:
Update COSMOS plugin helper script to exit with message if called with missing arguments

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+
+ ![`Commanding and Telemetry with OpenC3 Cosmos`](img/cosmos.png "Commanding and Telemetry with OpenC3 Cosmos")
+ 
+To best interact with the Linux assembly, we need to use a ground system interface, such as Cosmos. Before installing and running
```
Add hyperlink to the first "COSMOS" here to https://github.com/OpenC3/cosmos.

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+ ![`Commanding and Telemetry with OpenC3 Cosmos`](img/cosmos.png "Commanding and Telemetry with OpenC3 Cosmos")
+ 
+To best interact with the Linux assembly, we need to use a ground system interface, such as Cosmos. Before installing and running
+Cosmos we need to build the Cosmos plugin configuration files. This will allow Cosmos to decode telemetry from the Linux assembly and properly format
```
Cosmos should be instead COSMOS wherever you reference it. That is how OpenC3 spells it.

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,101 @@ Starting Linux demo... Use Ctrl+C to exit.
+Cosmos we need to build the Cosmos plugin configuration files. This will allow Cosmos to decode telemetry from the Linux assembly and properly format
+outgoing commands.
+
+These files are autocoded by Adamant. From the `adamant_example/src/assembly/linux/main` directory in your Adamant environment run:
```
Make it say "Adamant virtual environment" so it is very clear this is in the docker container.

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+
+ ![`Commanding and Telemetry with OpenC3 COSMOS`](img/cosmos.png "Commanding and Telemetry with OpenC3 COSMOS")
+ 
+To best interact with the Linux assembly, we need to use a ground system interface, such as OpenC3 COSMOS. To install COSMOS, navigate to an appropriate directory on your host machine and clone the COSMOS example project repository:
```
Add hyperlink to first mention of COSMOS to https://github.com/OpenC3/cosmos.

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+These files are autocoded by Adamant. From the `adamant_example/src/assembly/linux/main` directory in your Adamant environment run:
+
+```
+$ redo cosmos_config
````
To make sure a user is jumping between environments as little as possible, I think this redo command should come after all the openc3 plugin generation commands below. The redo should happen right before the user needs to do the copy.

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+This creates a new directory called `openc3-cosmos-linux-example`. Navigate to this directory and run:
+
+```
+$ ./../openc3.sh cli generate target LINUX_EXAMPLE
````
```
$ ../openc3.sh cli generate target LINUX_EXAMPLE
```
^ this is more standard

dinkelk replied:
I also think this would be clearer, instead of using "navigate" everywhere, just show the user:

```
$ cd openc3-cosmos-linux-example
$ ../openc3.sh cli generate target LINUX_EXAMPLE
```

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+cosmos-project/openc3-cosmos-linux-example/targets/LINUX_EXAMPLE/cmd_tlm/tlm.txt
+```
+
+Replace the contents of these text files with the contents of the corresponding files generated by Redo, located in the Adamant example project directory at:
````
`redo` instead of Redo

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+The plugin can now be compiled. Navigate to the plugin directory at `cosmos-project/openc3-cosmos-linux-example` and run:
+
+```
+$ ./../openc3.sh cli rake build VERSION=1.0.0
````
```
$ ../openc3.sh cli rake build VERSION=1.0.0
```

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+
+Select `Click to select plugin .gem file to install` and navigate to the compiled plugin gem file at `cosmos-project/openc3-cosmos-linux-example/openc3-cosmos-linux-example-1.0.0.gem` to install our generated plugin. The plugin is templated to allow changing of parameters such as ports, but the default values are already set to correspond with this project.
+
+The plugin will be installed. Return to the Adamant Linux environment and start the Linux assembly by navigating to `adamant_example/src/assembly/linux/main` and running:
```
Return to the Adamant virtual environment

dinkelk replied:
Again, instead of "navigating" just show:

```
$ cd adamant_example/src/assembly/linux/main
$ redo run
```

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+
+With COSMOS running, here are some interesting things you can try:
+
+ 1. View events generated from the Linux assembly in the main panel.
```
This is not yet possible. Or at least, these are not FSW generated events from Adamant. So maybe remove or reword?

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
````diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+redo run
+```
+
+You should see events being received by COSMOS every two seconds from the Linux assembly over the socket in the `CmdTLMServer` panel.
````
Maybe use the word "Log Messages", so this is not confused with FSW events, which is an Adamant concept.

dinkelk replied:
Also, I don't think this is actually happening. Nothing happens every two seconds. Maybe say that you should say "In the Log Messages panel you should see that the TCPIP server has accepted a new connection from Adamant once the assembly is up and running" or something like that

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+With COSMOS running, here are some interesting things you can try:
+
+ 1. View events generated from the Linux assembly in the main panel.
+ 2. View telemetry from the Counter and Oscillator components by opening the `Telemetry Grapher` panel and selecting the corresponding packets and parameters.
```
I think we need to be more specific here. Tell the user to select "LINUX_EXAMPLE" target, "HOUSEKEEPING_PACKET" packet and "OSCILLATOR_A.OSCILLATOR_VALUE.VALUE" as item and then click "Add Item". Just walk through specifically what to do.

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+
+ 1. View events generated from the Linux assembly in the main panel.
+ 2. View telemetry from the Counter and Oscillator components by opening the `Telemetry Grapher` panel and selecting the corresponding packets and parameters.
+ 3. Send any command by selecting it in the `Command Sender` panel. Try sending a NOOP or changing the Oscillator frequencies.
```
Same here.

dinkelk replied:
Also curious, everything in the command shows up as a modifiable "Parameter". I kind of like this, but is there a way to make some of the items unmodifiable?

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).
I haven't seen anything in the OpenC3 documentation that would set Parameters as un-modifiable, but there may still be some way that I'm not yet aware of.

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+ 1. View events generated from the Linux assembly in the main panel.
+ 2. View telemetry from the Counter and Oscillator components by opening the `Telemetry Grapher` panel and selecting the corresponding packets and parameters.
+ 3. Send any command by selecting it in the `Command Sender` panel. Try sending a NOOP or changing the Oscillator frequencies.
+ 4. View the queue usage for each component by opening the `Packet Viewer` panel and selecting the corresponding packet.
```
more specific here about what packet to select.

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

dinkelk commented on src/assembly/linux/main/README.md:
```diff
@@ -32,6 +32,108 @@ Starting Linux demo... Use Ctrl+C to exit.
+ 2. View telemetry from the Counter and Oscillator components by opening the `Telemetry Grapher` panel and selecting the corresponding packets and parameters.
+ 3. Send any command by selecting it in the `Command Sender` panel. Try sending a NOOP or changing the Oscillator frequencies.
+ 4. View the queue usage for each component by opening the `Packet Viewer` panel and selecting the corresponding packet.
+ 5. Send an interrupt to the running assembly by running `sh send_interrupt.sh` from a new SSH session within within the Linux environment. You should see the software respond by printing out `Interrupt received` with a time stamp in the terminal where `redo run` was started.
```
We don't SSH in anymore. Just say "from within another terminal in the Adamant virtual environment" and then show"

```
$ cd adamant_example/src/assembly/linux/main
$ sh send_interrupt.sh
```

Jbsco replied:
Addressed in [3d97fbf](https://github.com/lasp/adamant_example/pull/18/commits/3d97fbf738f7ddcddc9093b59cb591378c0e886d).

Jbsco committed:
Apply requested changes to COSMOS section of Readme

Apply additional changes to COSMOS section of Readme

Jbsco committed:
Update COSMOS helper script to provide script usage example on error

Jbsco committed:
Restructure COSMOS plugin file locations

Restructure COSMOS plugin configuration file locations, COSMOS plugin helper script changed to use paths relative to CWD, and README updated to encourage user to check plugin configuration template and use helper script.

Update COSMOS build script to capture new plugin location

Jbsco committed:
Add COSMOS instructions to Pico readme

Additional update to Pico readme

Jbsco committed:
Update Pico example readme to reflect COSMOS plugin fixes for serial interface

dinkelk committed:
Modify pico readme to use same serial device for all instructions

Jbsco committed:
Add convenience script to call the COSMOS plugin install script with its first argument from the main assembly directory

Make convenience scripts executable

Jbsco committed:
Restructure COSMOS plugin install script to force review/copy of templated plugin configuration

Jbsco committed:
Update COSMOS plugin install script to use `tr` for Unix compatibility

Jbsco committed:
Streamline the COSMOS plugin installation process

Fix COSMOS plugin naming for Linux example

Jbsco committed:
Update COSMOS Telemetry Grapher image for Pico example

dinkelk committed:
Add serial support for cosmos with a mac host

Since serial ports are not passable through to cosmos running on Docker
on a Mac, we provide a serial_tcp_bridge to forward serial traffic from
the Pico over TCP to COSMOS. This commit also wraps in some cleanup for
the installation scripts, and updates the READMEs appropriately.

dinkelk committed:
Add software status packet to pico assembly

dinkelk committed:
Add software status packet for linux assembly

Jbsco commented on src/assembly/pico/main/README.md:
```diff
@@ -171,35 +181,54 @@ cosmos-project/openc3-cosmos-pico-example/targets/PICO_EXAMPLE/cmd_tlm/cmd.txt
+This generates the command and telemetry configurations in the `adamant_example/src/assembly/pico/build/cosmos/plugin/` directory. Note that the main plugin configuration file is located in `adamant_example/src/assembly/pico/main/cosmos/plugin/`. This file should be reviewed to ensure the interface and required protocols are correct for your configuration. All of these files will be installed int ...
 
-The helper script, which takes the relative path from the top level of the assembly to the COSMOS install directory as an argument, copies the plugin configuration files, and any custom protocols used by the configuration, to the correct directories. If the COSMOS and Adamant example project directories are adjacent, complete the configuration by running:
+> **For Mac**, fist overwrite the provided `plugin.txt` file with one that is designed for TCP communication.
```
Likely "first", rather than "fist" - this wouldn't be picked up by `style`.
